evm/src/contracts/WormholeCctpTokenMessenger.sol

```diff
@@ -0,0 +1,346 @@
+     * @dev The CCTP message's source domain, target domain and nonce must match the VAA's.
+     * NOTE: This nonce is the one acting as the CCTP message sequence (and not the arbitrary one
+     * specified when publishing Wormhole messages).
+     */
+    error CctpVaaMismatch(uint32, uint32, uint64);
+
+    /**
+     * @dev Wormhole message finality. This value indicates a "finalized" consistency level, where
+     * finalized means the transaction where this message (event) lives will not be rolled back.
+     */
+    uint8 constant _MESSAGE_FINALITY = 1;
+
+    /**
+     * @dev Wormhole Core Bridge contract address.
+     */
+    IWormhole immutable _wormhole;
+
+    /**
+     * @dev Wormhole Chain ID. NOTE: This is NOT the EVM chain ID.
+     */
+    uint16 immutable _chainId;
+
+    /**
+     * @dev CCTP Message Transmitter contract interface.
+     */
+    IMessageTransmitter immutable _messageTransmitter;
+
+    /**
+     * @dev CCTP Token Messenger contract interface.
+     */
+    ITokenMessenger immutable _tokenMessenger;
+
+    /**
+     * @dev CCTP Token Minter contract interface.
+     */
+    ITokenMinter immutable _tokenMinter;
+
+    /**
+     * @dev CCTP domain for this network (configured by the CCTP Message Transmitter).
+     */
+    uint32 immutable _localCctpDomain;
+
+    constructor(address wormhole, address cctpTokenMessenger) {
+        _wormhole = IWormhole(wormhole);
+        _chainId = _wormhole.chainId();
+
+        _tokenMessenger = ITokenMessenger(cctpTokenMessenger);
+        _messageTransmitter = _tokenMessenger.localMessageTransmitter();
+        _tokenMinter = _tokenMessenger.localMinter();
+        _localCctpDomain = _messageTransmitter.localDomain();
+    }
+
+    /**
+     * @dev A convenience method to set the token spending allowance for the CCTP Token Messenger,
+     *  who will ultimately be burning the tokens.
+     */
+    function setTokenMessengerApproval(address token, uint256 amount) internal {
+        IERC20(token).approve(address(_tokenMessenger), amount);
+    }
+
+    /**
+     * @dev Method to burn tokens via CCTP Token Messenger and publish a Wormhole message associated
+     * with the CCTP Token Burn message. The Wormhole message encodes a `Deposit` (ID == 1), which
+     * has the same source domain, target domain and nonce as the CCTP Token Burn message.
+     *
+     * NOTE: This method does not protect against re-entrancy here because it relies on the CCTP
+     * Token Messenger to protect against any possible re-entrancy. We are leaning on the fact that
+     * the Token Messenger keeps track of its local tokens, which are the only tokens it allows to
+     * burn (and in turn, mint on another network).
+     */
+    function burnAndPublish(
+        bytes32 targetCaller,
+        uint32 targetCctpDomain,
+        address token,
+        uint256 amount,
+        bytes32 mintRecipient,
+        uint32 wormholeNonce,
+        bytes memory payload,
+        uint256 wormholeFee
```

Comment:
I'd prefer we don't add this constraint. If someone wants to use ETH funds in the integrating contract, he can be free to do so. Otherwise he can set a constraint like this in his contract prior to calling `burnAndPublish`. How does that sound?

Reply:
I'd prefer we don't add this constraint. If someone wants to use ETH funds in the integrating contract, he can be free to do so. Otherwise he can set a constraint like this in his contract prior to calling burnAndPublish. How does that sound?

cc: @gator-boi

Reply:
Yeah that's fine by me 👍 can we add a line to the docstring to document this intended design decision?

Reply:
Is this sufficient? https://github.com/wormhole-foundation/wormhole-circle-integration/pull/48/commits/8d4f00dbdd877ceeef7af07772bcf9d90eb49d1c

Reply:
Lgtm, thanks